rllib/evaluation/worker_set.py

```diff
@@ -601,48 +659,89 @@ def is_policy_to_train(
             raise NotImplementedError
 
     @DeveloperAPI
-    def foreach_worker(self, func: Callable[[RolloutWorker], T]) -> List[T]:
-        """Calls the given function with each worker instance as arg.
+    def foreach_worker(
+        self,
+        func: Callable[[RolloutWorker], T],
+        *,
+        local_worker=True,
+        # TODO(jungong) : switch to True once Algorithm is migrated.
+        healthy_only=False,
+        remote_worker_indices: List[int] = None,
+        timeout_seconds=None,
+    ) -> List[T]:
+        """Calls the given function with each worker instance as the argument.
 
         Args:
             func: The function to call for each worker (as only arg).
+            local_worker: Whether apply func on local worker too. Default is True.
+            healthy_only: Apply func on known active workers only. By default
+                this will apply func on all workers regardless of their states.
+            remote_worker_indices: Apply func on a selected set of remote workers.
+            timeout_seconds: Time to wait for results. Default is None.
 
         Returns:
              The list of return values of all calls to `func([worker])`.
         """
         local_result = []
-        if self.local_worker() is not None:
+        if local_worker and self.local_worker() is not None:
             local_result = [func(self.local_worker())]
-        remote_results = ray.get([w.apply.remote(func) for w in self.remote_workers()])
+
+        remote_results = self.__worker_manager.foreach_actor(
+            func,
+            healthy_only=healthy_only,
+            remote_actor_ids=remote_worker_indices,
+            timeout_seconds=timeout_seconds,
+        )
+        remote_results = [r.get() for r in remote_results.ignore_ray_errors()]
+
         return local_result + remote_results
 
     @DeveloperAPI
-    def foreach_worker_with_index(
-        self, func: Callable[[RolloutWorker, int], T]
-    ) -> List[T]:
-        """Calls `func` with each worker instance and worker idx as args.
+    def foreach_worker_async(
+        self,
+        func: Callable[[RolloutWorker], T],
+        *,
+        # TODO(jungong) : switch to True once Algorithm is migrated.
+        healthy_only=False,
+        remote_worker_indices: List[int] = None,
+    ) -> int:
+        """Calls the given function with each worker instance as the argument.
```

Comment:
Can you say something here on the async'ness of thie method vs the sync version of it?

Reply:
good catch. thanks.